tests/system/Debug/TimerTest.php

```diff
@@ -121,4 +123,107 @@ public function testReturnsNullGettingElapsedTimeOfNonTimer()
 
         $this->assertNull($timer->getElapsedTime('test1'));
     }
+
+    /**
+     * @timeLimit 1.5
+     */
+    public function testRecordFunctionNoReturn()
+    {
+        $timer       = new Timer();
+        $returnValue = $timer->record('longjohn', static function () { sleep(1); });
+
+        $this->assertGreaterThanOrEqual(1.0, $timer->getElapsedTime('longjohn'));
+        $this->assertNull($returnValue);
+    }
+
+    /**
+     * @timeLimit 1.5
+     */
+    public function testRecordFunctionWithReturn()
+    {
+        $timer       = new Timer();
+        $returnValue = $timer->record('longjohn', static function () {
+            sleep(1);
+
+            return 'test';
+        });
+
+        $this->assertGreaterThanOrEqual(1.0, $timer->getElapsedTime('longjohn'));
+        $this->assertSame('test', $returnValue);
+    }
+
+    public function testRecordArrowFunction()
+    {
+        $timer       = new Timer();
+        $returnValue = $timer->record('longjohn', static fn () => strlen('CI4'));
+
+        $this->assertLessThanOrEqual(1.0, $timer->getElapsedTime('longjohn'));
+        $this->assertSame(3, $returnValue);
+    }
+
+    public function testRecordThrowsException()
+    {
+        $this->expectException(RuntimeException::class);
+
+        $timer = new Timer();
+        $timer->record('ex', static function () { throw new RuntimeException(); });
```

Comment:
I have a question. If by mistake I made the callable throw an exception, then how can I get the elapsed time since `$this->stop($name)` will not be called? Should the invocation of the callable be wrapped in a `try-finally` ?

Reply:
If an exception is thrown and you don't catch it, then I think you might have other problems than worrying about the time stoppage.

But you also have a point. It does not handle the case of purposely catching exceptions for error handling.

```php
try {
    $benchmark->record('test', 'func_that_throws');
} catch (Throwable $e) {
    // stop() isn't called yet
}
```

@kenjis shall we cover this case as well?

Reply:
What if an exception is thrown after `$benchmark->start()` or in `$iterator->run()`?

Reply:
For `$iterator->run()`, nothing is caught internally. It would not record anything in the report and would go out of the method, ignoring all follow up Closures.

For `$benchmark->start()`, you would need to call `stop()` inside the `catch` block.

Reply:
I don't think we need `try-finally`. 

An exception should be thrown in an exceptional case, and not recoverable.
Normally, a developer would modify the code to prevent the exception from occurring.

If a developer really wants to know the elapsed time when an exception is thrown, they can catch and measure it by themselves.